`AsRef<Array<_, _>>` for `[T; N]` is poisonous

I recently attempted to update to `digest 0.11.0-rc.0` with _some_ of my crates. It appears having `hybrid-array` _anywhere_ in the dependency tree of a crate, not even in the immediate view, is sufficient for every existing `[T; N]::as_ref` call which prior worked to stop working _unless_ the caller explicitly specifies what they wanted a reference of as now there are multiple candidates for which `AsRef` is to be invoked.

I wouldn't raise this as an issue, except since having `hybrid-array` _anywhere_ as a dependency seems to trigger this, it causes any inclusion of `hybrid-array` in the dependency tree to require updating _all parts_ of a codebase. I have several crates, some decently distanced from the underlying hashes, which would require updating.

I wanted to flag this before the `digest 0.11` release was actually cut as hopefully, if this `impl AsRef` is removed, it won't require waiting for `digest 0.12`. Unfortunately, it would require a `hybrid-array 0.4.0`. With how pervasive this single `impl` is though, hence why my issue title calls it 'poisonous' as it poisons the entire tree, I'd hope to see this impl either removed or truly to be so beneficial, to be worth keeping. I don't have sufficient context to comment on the latter and concede that.

Aah, that's an interesting example. It shows the compiler will avoid the sized-to-unsized coercion entirely in the event the `AsRef` is ambiguous, even in a case where it can't select an appropriate `AsRef` impl whatsoever:

```
error[E0599]: the method `concat` exists for array `[&[u8; 6]; 2]`, but its trait bounds were not satisfied
 --> src/main.rs:3:37
  |
3 |   dbg!([b"abc".as_ref(), b"abcdef"].concat());
  |                                     ^^^^^^ method cannot be called on `[&[u8; 6]; 2]` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `[&[u8; 6]]: Concat<_>`

error[E0277]: the trait bound `[u8; 3]: AsRef<[u8; 6]>` is not satisfied
 --> src/main.rs:3:16
  |
3 |   dbg!([b"abc".as_ref(), b"abcdef"].concat());
  |                ^^^^^^ the trait `AsRef<[u8; 6]>` is not implemented for `[u8; 3]`
```

That is quite a wacky rule.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "hybrid-array"
-version = "0.3.1"
+version = "0.4.0-pre"
 description = """
 Hybrid typenum-based and const generic array types designed to provide the
 flexibility of typenum-based expressions while also allowing interoperability

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -156,32 +156,6 @@ pub type ArrayN<T, const N: usize> = Array<T, <[T; N] as AssocArraySize>::Size>;
 ///
 /// let arr: Array<u8, U3> = Array([1, 2, 3]);
 /// ```
-///
-/// ## [`AsRef`] impls
-///
-/// The [`AsRef`] trait can be used to convert from `&Array<T, U>` to `&[T; N]` and vice versa:
-///
-/// ```
-/// use hybrid_array::{Array, ArraySize, AssocArraySize, ArrayN, sizes::U3};
-///
-/// pub fn get_third_item_hybrid_array<T, U: ArraySize>(arr_ref: &Array<T, U>) -> &T {
-///     &arr_ref[2]
-/// }
-///
-/// pub fn get_third_item_const_generic<T, const N: usize>(arr_ref: &[T; N]) -> &T
-/// where
-///     [T; N]: AssocArraySize + AsRef<ArrayN<T, N>>
-/// {
-///     get_third_item_hybrid_array(arr_ref.as_ref())
-/// }
-///
-/// assert_eq!(get_third_item_const_generic(&[1u8, 2, 3, 4]), &3);
-/// ```
-///
-/// Note that the [`AssocArraySize`] trait can be used to determine the appropriate
-/// [`Array`] size for a given `[T; N]`, and the [`ArrayN`] trait (which internally uses
-/// [`AssocArraySize`]) can be used to determine the specific [`Array`] type for a given
-/// const generic size.
 #[repr(transparent)]
 pub struct Array<T, U: ArraySize>(pub U::ArrayType<T>);
 
@@ -486,17 +460,6 @@ where
     }
 }
 
-impl<T, U, const N: usize> AsRef<Array<T, U>> for [T; N]
-where
-    U: ArraySize<ArrayType<T> = [T; N]>,
-{
-    #[inline]
-    fn as_ref(&self) -> &Array<T, U> {
-        // SAFETY: `Self` is a `repr(transparent)` newtype for `[T; $len]`
-        unsafe { &*self.as_ptr().cast() }
-    }
-}
-
 impl<T, U> AsMut<[T]> for Array<T, U>
 where
     U: ArraySize,
@@ -514,17 +477,6 @@ where
     #[inline]
     fn as_mut(&mut self) -> &mut [T; N] {
         &mut self.0
-    }
-}
-
-impl<T, U, const N: usize> AsMut<Array<T, U>> for [T; N]
-where
-    U: ArraySize<ArrayType<T> = [T; N]>,
-{
-    #[inline]
-    fn as_mut(&mut self) -> &mut Array<T, U> {
-        // SAFETY: `Self` is a `repr(transparent)` newtype for `[T; $len]`
-        unsafe { &mut *self.as_mut_ptr().cast() }
     }
 }
 
@@ -663,7 +615,8 @@ where
 {
     #[inline]
     fn from(array_ref: &'a [T; N]) -> &'a Array<T, U> {
-        array_ref.as_ref()
+        // SAFETY: `Self` is a `repr(transparent)` newtype for `[T; $len]`
+        unsafe { &*array_ref.as_ptr().cast() }
     }
 }
 
@@ -683,7 +636,8 @@ where
 {
     #[inline]
     fn from(array_ref: &'a mut [T; N]) -> &'a mut Array<T, U> {
-        array_ref.as_mut()
+        // SAFETY: `Self` is a `repr(transparent)` newtype for `[T; $len]`
+        unsafe { &mut *array_ref.as_mut_ptr().cast() }
     }
 }
 

